index.bs

```diff
@@ -2781,6 +2792,54 @@ ScriptEvent = (
 
 ### Types ### {#module-script-types}
 
+#### The script.ExceptionDetails type #### {#type-script-Exception}
+
+[=Remote end definition=] and [=local end definition=]
+
+<pre class="cddl local-cddl remote-cddl">
+ExceptionDetails = {
+  columnNumber: uint,
+  exception: RemoteValue,
+  lineNumber: uint.
+  stackTrace: StackTrace
+  text: text,
+};
```

Comment:
I started implementing the `script.evaluate` proto, and figured out the fields `exception` and `stackTrace` are optional in the CDP. Was making them obligated intentional?
```suggestion
ExceptionDetails = {
  columnNumber: uint,
  exception?: RemoteValue,
  lineNumber: uint,
  stackTrace?: StackTrace,
  text: text,
};
```

Reply:
Do yo uhave an example of a script that would return an error with no `result` or `exceptionDetails.exception` value, or no `exceptionDetails.stackTrace`? Currently the prose part always populates those fields, so making them optional in the CDDL doesn't change anything.

Reply:
About **`exception `**: you are right. I can think only of the scenario when the script times out. But in this case the command failed on the protocol level:
```
{
  "id":9,
  "error":{
    "code":-32000,
    "message":"Execution was terminated"},
  "sessionId":"8DC93214FDAEE6DE8EC5D57AA35A0BE7"}
```

The **`stacktrace`** is added by the [`Debuger domain`](https://chromedevtools.github.io/devtools-protocol/tot/Debugger/), which is off by default, as it's quite expensive.

Reply:
```suggestion
ExceptionDetails = {
  columnNumber: uint,
  exception: RemoteValue,
  lineNumber: uint,
  stackTrace?: StackTrace,
  text: text,
};
```

Reply:
Hmm, so if one throws an actual exception object, that gets a `stack` property. So I don't know quite how lazilly that can be constructed, but it seems like in the case an actual exception is thrown (which I guess is the majority case vs just throwing a random object), the engine has to at least store enough information to allow constructing a stack anyway. So my inclination is to say we always have to provide a stack trace for an exception as part of the protocol.

Reply:
I'm not sure if I got the idea. Do you mean in case of throwing a random object BiDi should return an empty stack? Or treat the case somehow differently?

Reply:
No, I just mean that `throw new Error()` is going to return an object with a stack anyway. Given that, it seems totally reasonable to unconditionally put the stack into the response; it doesn't seem likely to actually have a significant extra perf cost in the engine given you're already generating a stack for most exceptions. For comparison, aiui Firefox's RDP always includes a stack when scipt evaluation results in an exception.

Reply:
I'm still not sure what should be the result of evaluating the following script. Could you please clarify?
```
const throwException = ()=>{
      throw {a:'a'};
  }
throwException();
```

Reply:
This code snippet generates a stacktrace in both Chrome or Firefox DevTools' console. Why should we have no stacktrace with BiDi? Is it also for performance reasons?

Or is it unclear because we should somehow make a difference between a script which fails to evaluate for something like a syntax error vs something that throws a custom error?

Reply:
Yes, I'd expect a stacktrace in the response. As @juliandescottes says it matches what you'd expect to happen in devtools. For Chrome in particular, evaluating that in the console gives a CDP response:
```
{
    "result": {
        "type": "object",
        "className": "Object",
        "description": "Object",
        "objectId": "5064291994948395578.1.3",
        "preview": {
            "type": "object",
            "description": "Object",
            "overflow": false,
            "properties": [
                {
                    "name": "a",
                    "type": "string",
                    "value": "a"
                }
            ]
        }
    },
    "exceptionDetails": {
        "exceptionId": 2,
        "text": "Uncaught",
        "lineNumber": 1,
        "columnNumber": 6,
        "scriptId": "34",
        "stackTrace": {
            "callFrames": [
                {
                    "functionName": "throwException",
                    "scriptId": "34",
                    "url": "",
                    "lineNumber": 1,
                    "columnNumber": 6
                },
                {
                    "functionName": "",
                    "scriptId": "34",
                    "url": "",
                    "lineNumber": 3,
                    "columnNumber": 0
                }
            ]
        },
        "exception": {
            "type": "object",
            "className": "Object",
            "description": "Object",
            "objectId": "5064291994948395578.1.4",
            "preview": {
                "type": "object",
                "description": "Object",
                "overflow": false,
                "properties": [
                    {
                        "name": "a",
                        "type": "string",
                        "value": "a"
                    }
                ]
            }
        }
    }
}
```
which clearly includes the stack. For BiDi I'd expect something like:
```
{
    "exceptionDetails": {
        "text": "Uncaught",
        "lineNumber": 1,
        "columnNumber": 6,
        "stackTrace": {
            "callFrames": [
                {
                    "functionName": "throwException",
                    "scriptId": "34",
                    "url": "",
                    "lineNumber": 1,
                    "columnNumber": 6
                },
                {
                    "functionName": "",
                    "scriptId": "34",
                    "url": "",
                    "lineNumber": 3,
                    "columnNumber": 0
                }
            ]
        },
        "exception": {
            "type": "object",
            "objectId": "5064291994948395578",
            "value": {
               "a": {
                   "type": "string",
                   "value": "a"
               }
           }
    }
}
```

Reply:
Indeed, if you make a described experiment in Chrome DevTools, you will get a stacktrace. But, getting into implementation details, it is because of the fact `Debugger` domain is enabled when DevTools is open. By default it is disabled, and no stacktrace will be provided.

Anyway, having or not a stacktrace in case of exception seems more a "debuggability" feature, so why not to keep it up to implementation?

Experiment to see difference between having `Debugger` domain enabled and disabled in Chrome:
0. Get a Chrome tab without an open DevTools.
1. Navigate to `data:text/html,<script>const a=()=>{throw 1;}; const b=()=>{a();};const c=()=>{b();};c();</script>`.
2. Open DevTools and go to console.
3. Verify there is no detailed stacktrace. Only `Uncaught 1`.
4. Reload the page with DevTools open.
5. Verify there exception is extendable with detailed stacktrace:
```
Uncaught 1
a @ ...
b @ ...
c @ ...
(anonymous) @ ...
```

Reply:
Leaving things up to the imlementation is usually not ideal because it means that consumers can't rely on a specific behaviour.

In this case, what's the argument from the user perspective for not providing a stacktrace for some exceptions? If my evaluated script fails and I don't expect it to, it seems like I'm always going to want a stacktrace to help debug the failure. Given that exception objects have a stack anyway, it doesn't seem like it can be too expensive to produce the stack in the js engine, so performance doesn't seem like a string argument from a user point of view.

Reply:
UPD: my bad. It's not a `Debugger` but a `Runtime` domain enriches CDP response with a stacktrace. Anyway, it still seems to be reasonable to keep it up for implementation.

Reply:
Ok, I agree. We can keep stackTrace obligatory.

```suggestion
ExceptionDetails = {
  columnNumber: uint,
  exception: RemoteValue,
  lineNumber: uint,
  stackTrace: StackTrace,
  text: text,
};
```